### PR TITLE
fix(api): count reconciliation backlog as work when deciding to idle-exit

### PR DESCRIPTION
**Pull request by jan-kubica**
## Proposed change

`createIdleExitCheck` decides batch mode is idle from
`countPendingDocumentProcessingJobs` alone, but that is only one of the two
sources of work the worker owns.

`initDocumentProcessingWorker` also runs `reconcileDocumentProcessing` on a
30s timer, and every phase there drains at most one batch per tick:
`RECONCILE_BATCH_SIZE` for delivery/repair/retry/stale-lease,
`SEARCH_INDEX_REPLAY_BATCH_SIZE` for reindex. None of that work passes through
the job queue, so a phase can return its whole batch (meaning more remained
than it could take) on the same tick the queue reads empty. The idle streak
keeps advancing, the worker exits mid-drain, and the remainder waits for
whatever starts the worker next, which has nothing to do with the backlog's
size.

The change:

- `RECONCILIATION_PHASE_BATCH_SIZE`, a total
  `Record<ReconciliationPhaseName, number>` beside the phase list, so a new
  phase cannot land without declaring its limit.
- `reconciliationLeftWorkBehind`: true when any phase returned its whole
  batch, or when a phase threw. A phase that threw keeps its count at 0,
  otherwise indistinguishable from drained.
- `createIdleExitCheck` now requires `hasUnfinishedReconciliation` and
  advances the streak only when the queue is empty *and* reconciliation found
  nothing left. Required rather than optional so a new call site must decide.

A tick that throws before producing counts sets the flag too: the backlog is
unknown, and this module's existing rule is to never exit on uncertainty, the
same reason a failed `countPending` resets the streak.

One trade-off worth an opinion: this also means a phase that fails
persistently holds the worker up instead of letting it exit. That is
consistent with how the module already treats a failing `countPending`, so I
kept it that way; if the billing side should win, the failure branch is the
line to change.

Note for merge order: #2104 touches the adjacent line in
`reconcileDocumentProcessing`. The two are independent and should merge
cleanly, but they land in the same function.

## Checklist

- [x] **BUILD ASSURANCE** | Code builds correctly and without any errors or warnings.
      `tsc --noEmit -p` on `apps/api` is clean, and scoped
      `oxlint -c oxlint.config.ts --deny-warnings --type-aware --type-check` on
      the touched files is clean.
- [x] **TESTING** | `bun test src/lib/document-processing-idle-exit.test.ts`
      (7 pass) and `bun test src/lib/document-processing-queue.test.ts`
      (48 pass). New cases cover: an empty queue is not idle while
      reconciliation has work behind it; reconciliation work mid-streak resets
      it; every declared phase is watched for saturation (asserted in both
      directions against the phase list); a partial batch is progress, not a
      backlog; a failed phase counts as work left behind. Verified locally
      because draft PRs skip repo CI.
- [ ] DARK MODE SUPPORT | Not applicable, no UI surface.
- [ ] ISSUE LINKED | No existing issue.
- [ ] SCREENSHOT INCLUDED | Not applicable, worker lifecycle only.

---
_Generated by [Claude Code](https://claude.ai/code/session_017yJi9krQ1b65jCU8z4jtuT)_

## Summary by CodeRabbit

- **Bug Fixes**
  - Workers no longer exit while document reconciliation has outstanding work.
  - Reconciliation activity correctly resets idle detection.
  - In-flight and failed reconciliation work is tracked to prevent premature shutdown.
  - Batch processing continues reliably when more work remains beyond the current limit.
  - Idle-exit sampling now starts safely and stops cleanly during shutdown.

- **Tests**
  - Added coverage for reconciliation progress, pagination, retries, cancellation, and safe worker exit behaviour.

### CODE REVIEW

**Comment by jan-kubica** [comment]
@codex review

File: apps/api/src/lib/document-processing-queue.ts [hunk 1]
```diff
@@ -2437,6 +2437,45 @@ type ReconciliationPhase = {
...
+export const reconciliationLeftWorkBehind = ({
+  counts,
+  failedPhases,
+}: {
+  counts: ReconciliationCounts;
+  failedPhases: readonly ReconciliationPhaseName[];
+}): boolean =>
+  failedPhases.length > 0 ||
+  RECONCILIATION_PHASE_NAMES.some(
+    (phase) => counts[phase] >= RECONCILIATION_PHASE_BATCH_SIZE[phase],
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/lib/document-processing-queue.ts]: Fixed in c2fd6cf: every phase now returns `{ count, hasMore }` and states saturation from its own knowledge, so no count is reinterpreted downstream.

- repair: the scan is cursor-paged, so `hasMore` is "the page came back full" (`resolveRepairScanPage`), which is the same fact as the resume cursor being non-null. The two are now derived together, so the cursor cannot advance while the phase reports drained. Its created-run count, which drops nearly every scanned field, no longer enters the decision.
- reindex / retry / stale-lease: `hasMore` is "the selection hit its `.limit()`", computed beside that limit. Their compare-and-set updates, claim losses, and failed replays reduce the count but say nothing about the backlog, and a failed replay is rescheduled with backoff, so it is not work the next tick could take.
- delivery: `dispatchQueuedDocumentProcessingRuns` reports `hasMore` from the rows it selected; `dispatchDocumentOcrBatches` now continues on that signal instead of re-deriving `attempted < limit` locally.
- a phase that threw reports `{ count: 0, hasMore: true }`: it drained nothing it can prove, which also removes the separate `failedPhases` list.

`reconciliationLeftWorkBehind` is now an OR over the declared phases only, so the batch-size mirror map is gone with the drift it carried.

Tests: the totality test asserts declared set equals exercised set in both directions over the phase names, flipping one phase's `hasMore` at a time; the synthetic partial-count case is replaced by `resolveRepairScanPage` cases at the real `RECONCILE_BATCH_SIZE` (a full page that created zero runs reports `hasMore` and advances the cursor; a short page ends the scan and resets it), plus a phase that returns a 10,000 count with `hasMore: false` reading as drained, a failed phase, and the scheduler mocks that now have to state `hasMore` to typecheck.

File: apps/api/src/lib/document-processing-queue.ts [hunk 2]
```diff
@@ -2437,6 +2437,45 @@ type ReconciliationPhase = {
...
+  );
+
+let unfinishedReconciliation = false;
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/lib/document-processing-queue.ts]: Fixed in c2fd6cf: the flag now tracks the tick, not just its result.

`createReconciliationProgress().runTick(tick)` sets `unfinished = true` in its synchronous prologue, before the first await, and assigns the tick's result only when the tick resolves. The sampler reads a plain boolean synchronously, so ordering is the whole contract: the reconcile loop calls `runTick` synchronously inside its detached start, and the assignment happens in that same turn. A tick that outlives an idle window, or that a sample lands inside, can no longer expose the previous tick's drained answer, and a rejected tick leaves the flag set because the assignment never runs. The ordering requirement is documented on both sides (the factory and the `hasUnfinishedReconciliation` option).

One consequence needed handling: both loops are intervals started at worker boot, and the 60s sample period is a multiple of the 30s reconciliation period, so unoffset samples would land on the same instant as every second tick. Since an in-flight tick now reads as unfinished, a phase-locked sampler would observe one at every sample and the worker would never exit, which is the opposite of what batch mode is for. Idle sampling therefore starts half a reconciliation period late (derived from `RECONCILE_INTERVAL_MS`, not a literal), putting samples between ticks; a tick still running by then has slow work genuinely in flight and should keep the worker alive.

Tests: an in-flight tick keeps `hasUnfinishedWork()` true synchronously and across later samples, then clears when it settles drained; a saturated tick keeps it set and a drained tick clears it; a rejected tick leaves it set; and the two halves are wired together in one test where a never-settling tick keeps `createIdleExitCheck` returning `checked` over an empty queue for the whole window and only exits after the tick reports drained. Also added an idle-exit case where reconciliation starts while the count is in flight, since that sample must not certify the moment before it.

**Comment by jan-kubica** [comment]
@codex review

File: apps/api/src/scripts/document-processing-worker.ts [hunk 3]
```diff
@@ -69,7 +83,9 @@ if (idleExitMinutes !== undefined) {
...
     },
   });
-  idleTimer = setInterval(() => {
-    detached(idleTick(), "document-processing.idle-exit-check");
-  }, IDLE_CHECK_INTERVAL_MS);
+  setTimeout(() => {
+    idleTimer = setInterval(() => {
+      detached(idleTick(), "document-processing.idle-exit-check");
+    }, IDLE_CHECK_INTERVAL_MS);
+  }, IDLE_CHECK_OFFSET_MS);
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/scripts/document-processing-worker.ts]: Fixed in a3b1b84. The deferred start is now owned by `startIdleExitSampling` (extracted next to `createIdleExitCheck`, the same seam the idle check itself uses), which returns a single `stop` handle covering both stages: it clears the pending `setTimeout` and any interval that already exists. The worker keeps that handle in `stopIdleSampling` and calls it from `shutdown` (right after `shuttingDown` flips, before the worker close) and from `onIdleExit`, which replaces the old local `idleTimer` clear. The deferred callback also re-checks `isShuttingDown()` before creating the interval, so a shutdown that raced the cancellation still cannot start one.

Regression tests in `document-processing-idle-exit.test.ts` drive the seam with a short offset: stopping inside the offset window records zero samples well past the offset and several intervals; a shutdown inside the window with no cancellation likewise records zero; and a positive control asserts sampling does start after the offset and stops for good once `stop` is called, so the two negative cases cannot pass vacuously.

File: apps/api/src/lib/scheduler/tasks/document-processing-ocr.test.ts [hunk 4]
```diff
@@ -9,7 +9,8 @@ test("drains queued OCR in bounded batches", async () => {
   const batchSizes = [100, 100, 17];
   const dispatch = mock(async ({ limit }: { limit?: number } = {}) => {
     expect(limit).toBe(100);
-    return { attempted: batchSizes.shift() ?? 0, retryAt: null };
+    const attempted = batchSizes.shift() ?? 0;
+    return { attempted, hasMore: attempted === limit, retryAt: null };
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/lib/scheduler/tasks/document-processing-ocr.test.ts]: Fixed in a3b1b84. The batching fixture now states `hasMore` per batch (`{ attempted: 100, hasMore: true }`, `{ attempted: 100, hasMore: true }`, `{ attempted: 17, hasMore: false }`) instead of deriving it from `attempted === limit`, which had re-encoded the count-based rule the loop no longer uses.

Added the case that separates the two rules: batches of `{ attempted: 12, hasMore: true }` then `{ attempted: 3, hasMore: false }` must dispatch 15 across two calls. Under the old `attempted < limit` condition the loop would stop after the first short batch, so this fails on a regression to it.

**Comment by jan-kubica** [comment]
@codex review

File: apps/api/src/lib/document-processing-queue.ts [hunk 5]
```diff
@@ -2472,53 +2618,60 @@ const reconcileDocumentProcessing = async ({
...
+          recoveredCount: String(
+            results[RECONCILIATION_PHASE.STALE_LEASE].count,
+          ),
+          reindexedCount: String(results[RECONCILIATION_PHASE.REINDEX].count),
+          repairedCount: String(results[RECONCILIATION_PHASE.REPAIR].count),
+          retriedCount: String(results[RECONCILIATION_PHASE.RETRY].count),
         });
-      },
+      }
+      return reconciliationLeftWorkBehind(results);
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/lib/document-processing-queue.ts:2839]: Valid, fixed in db7765a. The gap was real: the repair insert left `nextAttemptAt` null, and the reconciliation delivery phase selects scheduled work only, so nothing in the tick handed the new run to the queue.

Two changes close it:

- `persistMissingNativeExtractionRuns` now inserts the queued run with `nextAttemptAt: queuedAt` instead of leaving it unscheduled. The delivery phase runs last in the tick (repair first because it creates work, delivery last because it hands work off), so the run created by this tick is dispatched by this tick. If delivery cannot take them all, its own selection is saturated and it reports work behind, so the worker stays alive. The visible trade-off: `ASC NULLS FIRST` now sorts repaired runs behind never-dispatched ones, which matches their nature as backfill.
- `resolveRepairScanPage` treats created runs as a second, independent source of work left behind: the cursor still follows the scan alone (a short page resets it), while `hasMore` is `scanSaturated || createdRunCount > 0`. A page that ended the scan but created runs is therefore not drained.

Tests: `resolveRepairScanPage` gains the two-tick story as cases — a short page with created runs ends the scan (`nextCursor: null`) but reports `hasMore: true`, and the following short page that creates nothing reports drained. The phase-order invariant the delivery fix rests on is pinned separately: repair first, delivery last in the list the worker actually runs.

File: apps/api/src/scripts/document-processing-worker.ts [hunk 6]
```diff
@@ -2,12 +2,24 @@ import { envDocumentProcessingWorker } from "@/api/env-document-processing-worke
...
 
 const IDLE_CHECK_INTERVAL_MS = 60_000;
+// Both loops are intervals started at worker boot, and the sample period is
+// a multiple of the reconciliation period, so unoffset samples would land on
+// the same instant as every second reconciliation tick. A tick that is
+// running reads as unfinished by design, so a phase-locked sampler would see
+// one at every sample and never exit. Start half a reconciliation period
+// late, putting samples between ticks; a tick still running by then has slow
+// work genuinely in flight and should keep the worker alive.
+const IDLE_CHECK_OFFSET_MS = RECONCILE_INTERVAL_MS / 2;
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/scripts/document-processing-worker.ts]: Valid, and the offset is gone in db7765a rather than retuned: any fixed phase only separates the two cadences for ticks shorter than the gap it buys, so a 15-30s tick would have put every sample inside a run exactly as described.

The sample no longer races the tick. `hasUnfinishedReconciliation` is now async, and the progress tracker resolves it only once the tick in flight has reported: `runTick` publishes both "unfinished" and its in-flight promise before its first await, and a caller arriving mid-tick awaits that promise instead of reading a snapshot. The verdict a sample acts on is therefore always some tick's own result, never the previous tick's and never a race outcome, which makes it independent of alignment:

- slow but drained ticks overlapping every sample: the sample waits, reads drained, and the streak advances, so the worker exits;
- a continuously saturated loop: every tick reports work behind, so the streak never advances and the worker stays alive;
- a hung tick: the sample never completes, which is the same "never exit on uncertainty" rule the failed-count path already follows.

With the race gone the offset had no work left to do, so `IDLE_CHECK_OFFSET_MS` and the deferred-start helper are removed and the sampler is a plain interval again; the worker keeps a stop handle and clears it from both the shutdown and idle-exit paths.

Tests pin both alignments through `createIdleExitCheck` wired to a real tracker: a sample started inside a tick that ends saturated returns `checked` and does not exit even though the previous tick was drained, and a sample started inside a slow tick that ends drained returns `exit`.

File: apps/api/src/lib/document-processing-queue.test.ts [hunk 7]
```diff
@@ -1035,3 +1042,212 @@ describe("worker interruption lifecycle", () => {
...
+            name: "repair",
+            run: async () => ({ count: 10_000, hasMore: false }),
+          },
+        ]),
+      ),
+    ).toBe(false);
+  });
+
+  test("every declared phase states its own signal", async () => {
+    const drained = await resultsOf([]);
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/lib/document-processing-queue.test.ts:1125]: Valid, fixed in db7765a by binding the runnable list to the declaration and then testing that binding.

`RECONCILIATION_PHASE_RUNNERS` is `as const satisfies Record<ReconciliationPhaseName, ReconciliationPhase["run"]>`, so a declared phase without a producer fails typecheck. `DOCUMENT_PROCESSING_RECONCILIATION_PHASES` is assembled from that map over an explicit order (repair first, delivery last) and is what `reconcileDocumentProcessing` passes, so the only remaining way to leave a phase unrun is to omit it from that order.

The test now reads that exported list rather than only the initializer: declared names (the result record keys) equal the executed names (the production list), compared in both directions, before it flips each phase's signal one at a time. A phase declared but not wired in, or wired in but not declared, fails.

File: apps/api/src/lib/document-processing-idle-exit.test.ts [hunk 8]
```diff
@@ -97,3 +148,70 @@ describe("createIdleExitCheck", () => {
...
+
+  test("a shutdown inside the offset window never starts sampling", async () => {
+    const h = samplingHarness();
+
+    // Shutdown began but the deferred start was not cancelled: it must
+    // still refuse to create the interval.
+    h.beginShutdown();
+    await Bun.sleep(PAST_OFFSET_MS);
+
+    expect(h.samples()).toBe(0);
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/lib/document-processing-idle-exit.test.ts]: Resolved by removal in db7765a rather than by the suggested `try`/`finally`. The sampling offset these tests covered is gone: the idle sample now waits for a reconciliation tick in flight to report, which makes the verdict independent of timer alignment, so `startIdleExitSampling` and its deferred start no longer exist and the sampler is a plain interval again. With the helper gone the test and its leak window went with it.

File: apps/api/src/lib/document-processing-idle-exit.ts [hunk 9]
```diff
@@ -16,8 +29,48 @@ type CreateIdleExitCheckOptions = {
...
+  offsetMs,
+  onSample,
+}: StartIdleExitSamplingOptions): { stop: () => void } => {
+  let interval: ReturnType<typeof setInterval> | null = null;
+  const start = setTimeout(() => {
+    if (isShuttingDown()) {
+      return;
+    }
+    interval = setInterval(onSample, intervalMs);
+  }, offsetMs);
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/lib/document-processing-idle-exit.ts]: Correct reading of the code as it stood: the first sample landed at `offsetMs + intervalMs`, one idle-check interval later than the doc comment implied. Rather than fire a sample at the offset, db7765a removes the offset entirely. It existed only to keep samples off the reconciliation tick boundary, and a fixed phase cannot do that for ticks longer than the offset; the sample now waits for a tick in flight to report instead, which is alignment-independent. Sampling is a plain interval again, so the first sample is at one interval, as before this PR.

**Comment by jan-kubica** [comment]
@codex review

File: apps/api/src/lib/document-processing-idle-exit.ts [hunk 10]
```diff
@@ -36,7 +51,12 @@ export const createIdleExitCheck = ({
     inFlight = true;
     try {
       const pending = await countPending();
-      consecutiveIdleChecks = pending === 0 ? consecutiveIdleChecks + 1 : 0;
+      // Read reconciliation after the count settles, and wait for a tick in
+      // flight: a tick that starts while the count runs still lands in this
+      // sample, and one that outlives the count is answered by its own
+      // verdict rather than by whichever finished first.
+      const idle = pending === 0 && !(await hasUnfinishedReconciliation());
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/lib/document-processing-idle-exit.ts]: Valid, fixed in 289f77f by ordering rather than more state.

The sample now settles reconciliation first and counts second. Reconciliation is a producer of queue entries and never consumes the sample, so waiting for the tick in flight to report before counting means everything that tick enqueued is already in the count; the old direction paired one tick's pre-enqueue count with that same tick's drained verdict, which is exactly the window where the delivery phase hands off recovered runs. The in-code comment now argues this direction instead of the reverse, and states the follow-on invariant: a tick that starts after the verdict has marked itself unfinished before its own first await, so the next sample reads it as running rather than reading its result, and no idle streak can span it.

Tests:

- a tick in flight whose queue depth goes 0 to 3 before it reports drained: the sample returns `checked` and does not exit. Reverting the two reads to the previous order turns that into `exit`, so the case is discriminating rather than decorative.
- the ordinary drained case still exits: the existing slow-tick test keeps an empty queue throughout and returns `exit`.
- the "starts during the sample" case was retargeted to the new invariant: a tick that begins while the count runs is caught by the following sample, which resets the streak, so two consecutive idle checks cannot span it.

**Comment by jan-kubica** [comment]
@codex review

File: apps/api/src/lib/document-processing-idle-exit.ts [hunk 11]
```diff
@@ -35,8 +50,19 @@ export const createIdleExitCheck = ({
...
+      // already in the count. Counting first would pair a pre-enqueue
+      // count with the same tick's drained verdict and read idle while
+      // fresh jobs wait, which a worker close does not drain. A tick that
+      // starts after this verdict has already marked itself unfinished
+      // before its own first await, so the next sample reads it as
+      // running rather than reading its result.
+      const unfinished = await hasUnfinishedReconciliation();
       const pending = await countPending();
-      consecutiveIdleChecks = pending === 0 ? consecutiveIdleChecks + 1 : 0;
+      const idle = pending === 0 && !unfinished;
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/lib/document-processing-idle-exit.ts]: Valid, fixed in 5d89cbd with a synchronous re-read rather than another await.

The progress tracker now also answers `isTickRunning()` synchronously, set in the tick's own prologue before its first await, alongside the awaited verdict. A sample resolves the verdict, awaits the count, and then, in the frame the count resumes it in, re-reads that flag and decides:

```
const unfinished = await hasUnfinishedReconciliation();
const pending = await countPending();
const idle = pending === 0 && !unfinished && !isReconciliationInFlight();
```

No await separates that read from the streak update and the `onIdleExit` call, and neither timers nor microtasks interleave inside a frame, so any tick that began before the decision is visible to it and none can begin during it. The window this closes is exactly the mirror of the previous one: a tick starting after the verdict and before the count returns, on the final sample a streak needs, where no later sample exists to catch it. What the exit callback then does with its own scheduling is the callback's concern; the decision itself is race-free.

Tests, both mutation-checked (dropping the re-read makes both fail, dropping it leaves the round-four ordering test green, so they pin distinct windows):

- unit: `countPending` resolves through a `then` handler that marks reconciliation running, with `requiredIdleChecks: 1` so nothing later could compensate. The sample returns `checked`.
- integration against the real tracker: the same handler starts an actual `runTick`, so the flag comes from the production prologue rather than a stand-in. The sample returns `checked`; once the tick settles drained, the next sample returns `exit`.
- the ordinary case is pinned in both files: a final sample with nothing running still exits.

**Comment by jan-kubica** [comment]
@codex review

File: apps/api/src/lib/document-processing-idle-exit.ts [hunk 12]
```diff
@@ -35,8 +58,25 @@ export const createIdleExitCheck = ({
...
+      // The mirror of that window: a tick can start after the verdict and
+      // while the count is in flight, and on the last sample of a streak
+      // there is no later sample to catch it. So re-read the running flag
+      // here and decide in this frame. Reconciliation sets that flag in
+      // its tick's synchronous prologue, before its own first await, and
+      // nothing else runs between this read and the exit call below: no
+      // await separates them, and neither timers nor microtasks interleave
+      // inside one frame. Every tick that began before the decision is
+      // therefore visible to it, and none can begin during it.
+      const idle = pending === 0 && !unfinished && !isReconciliationInFlight();
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/lib/document-processing-idle-exit.ts]: Valid, fixed in 8c42d89 with a generation counter, which closes the family rather than this one member.

The tracker increments a tick counter in the same synchronous prologue that already sets the running flag. The sample snapshots it after the verdict, counts, and compares in the frame it decides in:

```
const unfinished = await hasUnfinishedReconciliation();
const generation = reconciliationGeneration();
const pending = await countPending();
const idle =
  pending === 0 &&
  !unfinished &&
  !isReconciliationInFlight() &&
  reconciliationGeneration() === generation;
```

Any tick that begins anywhere in the count window moves the counter, including the one that also ends there and so leaves neither a running flag nor a verdict this sample ever read. That makes the enumeration total, and the comment says so: a tick either began before the verdict this sample waited on (`unfinished`, `pending`), or inside the count and is still running (`isReconciliationInFlight`), or inside the count and finished (the generation), or after the decision, which one frame makes impossible. Each term covers a case the others do not.

A moved generation is not proof of work, so the sample is deliberately conservative there and the next sample reads that tick's own verdict, which is why both halves are pinned:

- saturated tick that fits inside the count: the sample returns `checked`, and so does the one after it, on that tick's verdict.
- drained tick that fits inside the count: the sample returns `checked` even though every other reading says drained, and the next sample returns `exit`, so conservatism costs one check rather than the exit.
- unit-level equivalent with stubs, plus the existing rounds four and five cases and the plain drained final sample, all still green.

Mutation-checked: dropping the generation term fails exactly those three tests and nothing else, so it is not covered by the earlier terms.

**Comment by jan-kubica** [comment]
@codex review

File: apps/api/src/lib/document-processing-queue.ts [hunk 13]
```diff
@@ -2402,7 +2499,10 @@ const recoverStaleDocumentProcessingRuns = async (): Promise<number> => {
       ),
     )
     .returning({ id: documentProcessingRuns.id });
-  return exhausted.length + searchReplays.length + recovered.length;
+  return {
+    count: exhausted.length + searchReplays.length + recovered.length,
+    hasMore,
+  };
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/lib/document-processing-queue.ts:2567]: Valid, and fixed in a4c5ba1 as a class rather than as this edge.

**The DAG.** Reading each phase's writes against every phase's selection predicate, the same-tick edges are:

- repair inserts queued runs due now (delivery selects queued rows with a due schedule) and, where a current projection exists, failed runs coded `search_index_failed` due now (reindex selects those): repair feeds reindex, delivery.
- retry moves retryable failures to queued due now: retry feeds delivery.
- stale-lease returns expired leases to queued due now (delivery) and expired search-index replays to failed, `search_index_failed`, unscheduled, which reindex selects: stale-lease feeds reindex, delivery. This is the finding. Its third write, exhausted attempts coded `worker_lease_expired` past the attempt cap, matches no phase's predicate (that code is not in the retryable set, and the cap excludes it anyway).
- reindex claims rows for itself; completions are terminal and non-completions come back to reindex with a backoff, so it feeds no other phase.
- delivery only moves rows into the job queue or reschedules them for a later delivery.

That is acyclic, so ordering can absorb all of it: repair, retry, stale-lease, reindex, delivery. The edges are declared next to the order as a total map, and a test binds them in both directions (every declared phase appears in the order, and every producer precedes every phase it feeds). Reverting the order to what it was fails exactly that test.

**What ordering cannot place.** Two transitions return work to the phase that produced it, so no order helps and the phase reports them itself:

- reindex: a replay it claimed and did not complete is back on `failed` with a backoff, which is its own input. `resolveSearchIndexReplayBatch` reports `recoveredCount < claimedCount`.
- delivery: a handoff that failed leaves the run queued and rescheduled, and delivery is the only phase that takes queued rows. `dispatchScheduledDocumentProcessingRetries` now reports `retryAt !== null`.

Both are pinned by their own tests, and dropping either term fails only that test. No other unlinearizable edge remains.

**One term removed.** Repair's created-run term is gone: its output is due input for reindex and delivery, which now both run later in the same tick and answer for it through their own saturation and the two terms above. That closes the round-three edge more completely than the term did, since it covers runs created by retry and stale-lease too, which the repair-specific term never reached.

Limit worth naming: the edge map is a reading of the predicates, so it is checked against the order but not against the queries themselves. A wrong or missing edge there would need a database-level test, which this suite has none of.

**Comment by jan-kubica** [comment]
@codex review

File: apps/api/src/lib/document-processing-idle-exit.ts [hunk 14]
```diff
@@ -35,8 +66,36 @@ export const createIdleExitCheck = ({
...
+      // finished inside the count, leaving neither a running flag nor a
+      // verdict this sample ever read, having possibly enqueued after the
+      // count snapshot; a moved generation is not proof of work, so this
+      // sample is conservative and the next one reads that tick's own
+      // verdict. Every term covers a case the others provably do not.
+      const idle =
+        pending === 0 &&
+        !unfinished &&
+        !isReconciliationInFlight() &&
+        reconciliationGeneration() === generation;
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/lib/document-processing-idle-exit.ts]: Valid, fixed in 5ae512b by making the crossing mean "not finished measuring" rather than "not idle".

A sample that finds a tick crossed its readings now measures again: it awaits the verdict of the tick that crossed it, re-counts what that tick may have enqueued, and re-checks in the frame it decides in. The reads are unchanged; only the response to a crossing changed.

- drained system: the second pass reads that tick's own verdict, which is drained, and the count it takes is after whatever the tick enqueued, so the sample completes. The alignment you describe now exits within the streak instead of never.
- busy system: the verdict is unfinished and the sample reads not-idle from the same place it always did.
- pathological producer: the chase is bounded at three, after which the sample is conceded as not idle. A producer that starts a tick inside every count is not something to certify as drained, however drained each tick claims to be.

The comment now says that the generation term keeps its job of noticing a tick that left no other trace, and that noticing means the measurement is unfinished rather than the sample is spent.

Tests, all mutation-checked (disabling the chase fails exactly these three):

- every sample crossed by a drained tick, two idle checks required: exits on the second sample.
- a sample crossed by a saturated tick: reads that tick's verdict and does not exit.
- a producer crossing every count: conceded as not idle, having measured more than once and fewer than eight times, so it neither decided on the first reading nor spun.

Two round-six expectations changed deliberately. A drained tick that fits inside the count now resolves within the same sample rather than costing one, which is the correct outcome under this model and is pinned as such; a tick that starts as the count resolves is now waited out by that sample, so its test settles it saturated and asserts the sample does not exit.

File: apps/api/src/lib/document-processing-queue.test.ts [hunk 15]
```diff
@@ -1035,3 +1046,410 @@ describe("worker interruption lifecycle", () => {
...
+    ).toBe(true);
+  });
+
+  test("every phase runs ahead of the phases it feeds", () => {
+    const order = DOCUMENT_PROCESSING_RECONCILIATION_PHASES.map(
+      ({ name }): string => name,
+    );
+    const feeds = Object.entries(
+      DOCUMENT_PROCESSING_RECONCILIATION_PHASE_FEEDS,
+    );
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/lib/document-processing-queue.test.ts:1167]: Fair, and done in c8db428: the edge map is now checked against the database rather than against a reading of the predicates.

`document-processing-reconciliation-feeds.db.test.ts` drives the real phases through the exported phase list, against the test database with `@/api/db/root` pointed at it. For each producer-consumer pair it resets state, seeds the producer's input, runs the producer, records the rows it wrote, runs the consumer, and records whether the consumer took any of them. Then:

- every phase must have written something, so no fixture is silently inert;
- observed edges must equal declared edges, both directions in one assertion;
- the order the worker runs must place every producer ahead of every phase it feeds.

Seven fixtures cover every transition the phases make: repair with and without an existing projection (a queued run due now, and a failed `search_index_failed` run), retry requeueing a retryable failure, stale-lease on an expired lease and on an expired search-index replay, reindex replaying a failed index, delivery handing a scheduled run to the queue. Observed comes out exactly equal to declared: repair feeds reindex and delivery, retry feeds delivery, stale-lease feeds reindex and delivery. Removing `stale-lease -> reindex` from the map, the edge this thread reported, now fails the test with that edge listed as observed but undeclared.

Redis, the queue producer, the search provider, and realtime are stubbed. None of them decide which rows a phase selects, which is the only thing observed here; the transitions and predicates are the real ones. No edge is left reading-verified.

**Comment by jan-kubica** [comment]
@codex review

**Comment by jan-kubica** [comment]
@codex review

**Comment by jan-kubica** [comment]
@codex review
